packages/artifact/src/internal-artifact-client.ts

```diff
@@ -100,25 +133,113 @@ export class DefaultArtifactClient implements ArtifactClient {
     return uploadResponse
   }
 
-  /*
-  Downloads a single artifact associated with a run
+  async downloadArtifact(
+    name: string,
+    path?: string | undefined,
+    options?: DownloadOptions | undefined
+  ): Promise<DownloadResponse> {
+    const artifacts = await listArtifacts()
+    if (artifacts.count === 0) {
+      throw new Error(
+        `Unable to find any artifacts for the associated workflow`
+      )
+    }
+
+    const artifactToDownload = artifacts.value.find(artifact => {
+      return artifact.name === name
+    })
 
-  export async function downloadArtifact(
-      name: string,
-      path?: string,
-      options?: DownloadOptions
-    ): Promise<DownloadResponse> {
+    if (!artifactToDownload) {
+      throw new Error(`Unable to find an artifact with the name: ${name}`)
+    }
+    // query container for all items associated with the artifact
+    const items = await getContainerItemsForSingleArtifact(
+      artifactToDownload.name,
+      artifactToDownload.fileContainerResourceUrl
+    )
+
+    if (!path) {
+      path = getWorkSpaceDirectory()
+    }
 
-    TODO
+    // During upload, empty directories are rejected by the remote server so there should be no artifacts that consist of only empty directories
+    const downloadSpecification = getDownloadSpecification(
+      name,
+      items.value,
+      path,
+      options?.createArtifactFolder || false
+    )
+
+    if (downloadSpecification.filesToDownload.length === 0) {
+      core.info(
+        `No downloadable files were found for the artifact: ${artifactToDownload.name}`
+      )
+    } else {
+      // Create all necessary directories recursively before starting any download
+      await createDirectoriesForArtifact(
+        downloadSpecification.directoryStructure
+      )
+      await downloadSingleArtifact(downloadSpecification.filesToDownload)
+    }
+
+    return {
+      artifactName: name,
+      downloadPath: downloadSpecification.rootDownloadLocation
+    }
   }
 
-  Downloads all artifacts associated with a run. Because there are multiple artifacts being downloaded, a folder will be created for each one in the specified or default directory
+  async downloadAllArtifacts(
+    path?: string | undefined
+  ): Promise<DownloadResponse[]> {
+    const response: DownloadResponse[] = []
+    const artifacts = await listArtifacts()
+    if (artifacts.count === 0) {
+      core.info('Unable to find any artifacts for the associated workflow')
+      return response
+    }
 
-  export async function downloadAllArtifacts(
-      path?: string
-    ): Promise<DownloadResponse[]>{
+    if (!path) {
+      path = getWorkSpaceDirectory()
+    }
+    path = normalize(path)
+    path = resolve(path)
 
-      TODO
+    // Get container items for all artifacts. The resource url is the same for all artifacts so one call is sufficient
+    const items = await getContainerItemsForAllArtifacts(
+      artifacts.value[0].fileContainerResourceUrl
```

Comment:
The container should have a `containerId` field, we should make one call per `containerId`.

Reply:
The principle is we should not assume any server-side implementation details if possible.

Reply:
So `fileContainerResourceUrl` is effectively just `${runtimeUrl}_apis/resources/Containers/${containerId}`.  Both the `containerId` and the `fileContainerResourceUrl` are the same for all the artifacts since the same container is shared.

Regardless if we use `containerId` + `runtimeUrl` or just the `fileContainerResourceUrl` URL, in both cases we need to know that artifacts for the same run share the same container

Reply:
Did some changes to make one call per artifact with each `fileContainerResourceUrl`. Removes the assumptions related to how the artifacts get stored